### PR TITLE
API Cleanup

### PR DESCRIPTION
**Pull request by AlexDBlack**
Good to go IMO. I'll leave this open a bit longer for reviews.

Addresses:
- https://github.com/deeplearning4j/deeplearning4j/issues/3889 (Fixed override, added proposed convenience method: ```.inputType().convolutional(h,w,d)```)
- https://github.com/deeplearning4j/deeplearning4j/issues/3888 (.layer(Layer) overload now exists, in addition to .layer(int, Layer))
- https://github.com/deeplearning4j/deeplearning4j/issues/3897 (.regularization(boolean) deprecated, use removed internally)
- https://github.com/deeplearning4j/deeplearning4j/issues/3896 (improved dropout javadoc)
- https://github.com/deeplearning4j/deeplearning4j/issues/3895 (.units(int) is equivalent to .nOut(int))

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 1]
```diff
@@ -213,6 +214,7 @@ public double getL2ByParam(String variable) {
      * Fluent interface for building a list of configurations
      */
     public static class ListBuilder extends MultiLayerConfiguration.Builder {
+        private AtomicInteger layerCounter = new AtomicInteger(-1); //Used only for .layer(Layer) method
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: AtomicInteger sounds like ("communicates") you're expecting concurrency, while that is not true - the use of an ordinary map rules out concurrency. You could thus just use an int. 

However, I think it would be wiser to just use the "find max key in map" logic and then just invoke the indexed variant with max+1, as that is very straightforward and immediately understandable for a code-reader, and there is no possibility to start wondering what happens if you mix the methods. Right now, while reviewing this, I am wondering what happens if I e.g. add layers using index in out-of-order fashion, and then add one layer using the non-indexed method. After thinking a bit, I can't immediately see how that would fail - but my point is that this particular thinking should really not be necessary!

Also, JavaDoc, explaining what they do, and how they interact.

I would note that the "same index overrides" has tripped me up over a dozen times already, and I am still just getting to know the API! It should rather have thrown - it makes NO sense that this is possible, IMHO. It should at least be JavaDocced, and could potentially also output a log.warn(..)

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 2]
```diff
@@ -261,6 +262,19 @@ public ListBuilder layer(Layer layer){
             return layerwise;
         }
 
+        @Override
+        public ListBuilder setInputType(InputType inputType){
+            return (ListBuilder)super.setInputType(inputType);
+        }
+
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Since my ML-level hasn't reached anything else than feedforward at this point, this might be utterly meaningless: Are you sure it makes sense for a ListBuilder to accept all the different types of InputType? Will the ListBuilder - in this usage scenario - be used in a non-feedforward setting? I had this idea that these other types would be used "within" the network, but not as the top, "receiving" part? As I said, might be totally off base here.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 3]
```diff
@@ -1008,17 +1008,23 @@ public Builder l2Bias(double l2Bias) {
...
+         * Note 1: This sets the probability per-layer. Care should be taken when setting lower values for
+         * complex networks (too much information may be lost with aggressive dropout values).<br>
+         * Note 2: Frequently, dropout is not applied to input (first layer) our output layer. This needs to be
+         * handled MANUALLY by the user - set .dropout(0) on those layers when using global dropout setting.<br>
+         * Note 3: Implementation detail (most users can ignore): DL4J uses inverted dropout, as described here:
+         * <a href="http://cs231n.github.io/neural-networks-2/">http://cs231n.github.io/neural-networks-2/</a>
          * </p>
          *
-         * @param dropOut Dropout probability (probability of retaining an activation)
+         * @param inputRetainFraction Dropout probability (probability of retaining an input activation for a layer)
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Point out (again) here in the @param section that 0.0 _disables_ dropout, while 1.0 _effectively disables_ it, since it means that all nodes are retained.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 2 above)
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: e.g. Embedding would be another common type

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 4]
```diff
@@ -1008,17 +1008,23 @@ public Builder l2Bias(double l2Bias) {
...
+         * Dropout probability. This is the probability of <it>retaining</it> an input activation for a layer. So
+         * dropOut(x) will keep an input activation with probability x, and set to 0 with probability 1-x.<br>
+         * dropOut(0.0) is disabled (default). When {@link #useDropConnect(boolean)} is set to true (false by default),
+         * this method sets the drop connect probability instead.
          * <p>
-         * Note: This sets the probability per-layer. Care should be taken when setting lower values for complex networks.
+         * Note 1: This sets the probability per-layer. Care should be taken when setting lower values for
+         * complex networks (too much information may be lost with aggressive dropout values).<br>
+         * Note 2: Frequently, dropout is not applied to input (first layer) our output layer. This needs to be
+         * handled MANUALLY by the user - set .dropout(0) on those layers when using global dropout setting.<br>
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Point again out that 0 disables it. Point is that this disable-value is hard to grasp, as it semantically means "keep NO nodes"..

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 5]
```diff
@@ -1008,17 +1008,23 @@ public Builder l2Bias(double l2Bias) {
         }
 
         /**
-         * Dropout probability. This is the probability of <it>retaining</it> an activation. So dropOut(x) will keep an
-         * activation with probability x, and set to 0 with probability 1-x.<br>
-         * dropOut(0.0) is disabled (default).
+         * Dropout probability. This is the probability of <it>retaining</it> an input activation for a layer. So
+         * dropOut(x) will keep an input activation with probability x, and set to 0 with probability 1-x.<br>
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: I am not sure whether I understand this. Because if I do, I think it is wrong.
Dropout should "zero out _nodes_", and I've come to understand that this means that their OUTPUT will be zero. Which again means that the NEXT layer will see a consistent set of zeroed and non-zeroed node values, i.e. if for Layer L node 0, 2 and 5 is zeroed, then ALL nodes in Layer L+1 will see node 0, 2 and 5 as zeroed.

Reading this JavaDoc, I get the feeling that node 0 in Layer L+1 (which is THIS layer, since we're talking "input activations") might see that the input activation is zeroed for node 1, 2, 3 from Layer L, while node 1 in Layer L+1 could see zeroed nodes for 0, 2 and 3 - i.e. independently zeroed.

If the latter is the case, then I believe this dropout is implemented incorrectly. But I probably misunderstand the wording.

Also, with dropout, one should have this scaling thing. This should be mentioned in the JavaDoc. Here's TensorFlow's dropout doc:
https://www.tensorflow.org/api_docs/python/tf/nn/dropout

"With probability keep_prob, outputs the input element scaled up by 1 / keep_prob, otherwise outputs 0. The scaling is so that the expected sum is unchanged."

Which brings us to the next thing that should be javadocced, and that is that dropout should only be applied in the training phase, not the evaluation. In particular, this also goes for the scaling. Since when using TensorFlow, you as user needs to set the keep_prob to your wanted value when training, while in evaluation mode, **you need to set it to 1**. Using e.g. "0.5", this means that when training, half of the nodes are zeroed, but the output is scaled by 2 for the remaining. But in evaluation, no nodes are silenced, and there is no scaling.

Is this handled automatically by sl4j? Would be nice to have it here in the JavaDoc.

Regarding whether it is "before" or "after" this layer: This could be explicitly explained like "setting a dropout on this layer is the same as adding a {@link DropoutLayer} _in front_ of this layer".

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 6]
```diff
@@ -1008,17 +1008,23 @@ public Builder l2Bias(double l2Bias) {
...
+         * dropOut(0.0) is disabled (default). When {@link #useDropConnect(boolean)} is set to true (false by default),
```
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Something like "dropout(0) is a special value that _disables_ the dropout mechanism, which effectively is the same as dropout(1.0) since that means to retain all nodes."

Problem is that 0 _semantically_ would mean to drop _all_ nodes, which can hinder understanding unless being very explicit about it.

Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: keras' definition is not exactly cleaner btw: 

rate: float between 0 and 1. Fraction of the input units to drop.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 4 above)
Comment (stolsvik) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Regarding input layer. The "dropout paper" (_the original!_) suggests 0.8 retain on input, 0.5 on hidden layers. So yes, it should definitely be different, but not necessarily "retain all".

AGAIN the point about "0": I consistently get a "huh?!"-feeling seeing that number - which implies that it should always be pointed out something like _".. thus disabling the dropout for this layer"_, or something.

Yes, I realize that I really don't like that special value, as it so directly contradicts the semantic of what the number means.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 1 above)
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Changed to int.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 2 above)
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Yes, it does make sense to accept all input types.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 3 above)
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Additional javadoc changes + changed to inputRetainProbability

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java (hunk 5 above)
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Dropout zeros out activations, by definition.
The "scaling thing" is mentioned in the inverted dropout implementation.
Added note on train vs. test.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java [hunk 7]
```diff
@@ -1008,17 +1008,23 @@ public Builder l2Bias(double l2Bias) {
...
+         * this method sets the drop connect probability instead.
          * <p>
-         * Note: This sets the probability per-layer. Care should be taken when setting lower values for complex networks.
+         * Note 1: This sets the probability per-layer. Care should be taken when setting lower values for
+         * complex networks (too much information may be lost with aggressive dropout values).<br>
+         * Note 2: Frequently, dropout is not applied to input (first layer) our output layer. This needs to be
```
Comment (sorensf) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java]: Note 2: Frequently, dropout is not applied to input (first layer) _our_ output layer.

"our" should probably be "or". And since the java doc is the same for Layer.java any changes here should be reflected there as well.
